### ISSUE TITLE
Password revocery feature - new password is not working

### ISSUE THREAD

**Issue by BujakiAttila** [JOBAA/amos-ss15-proj4#68]
By the 'Forgot Password' feature generated password is not working. The email sent with the generated password with no problem, but the generated password is not accepted at authentication. 

ClientController.java - ClientForgotPassword method generates the password hash:

`String newPasswordHash = new BCryptPasswordEncoder().encode(newRandomPassword);`

This is not creating a valid hash. When trying to log in on the user interface using the "newRandomPassword", the credentials are not accepted. Maybe salting is not done properly? Any ideas?

**Comment by JonasGroeger** [comment]
There is no such thing as manual salting in place right now. BCrypt itself already is already doing that. See http://stackoverflow.com/questions/6832445/how-can-bcrypt-have-built-in-salts

**Comment by BujakiAttila** [comment]
Yes, the encofdng "strength" should be passed to the BCryptPasswordEncoder constructor. It should be easy to fix.

*About my problem:*
It was just an idea that the salting is the cause. I am updating the DB content. Maybe the credentials info is loaded to the memory, so it would be faster. In other examples I have seen them using the "UserService".

Like this:
```java
public GenericResponse savePassword(Locale locale, @RequestParam("password") String password) {
    User user = (User) SecurityContextHolder.getContext().getAuthentication().getPrincipal();
    userService.changeUserPassword(user, password);
    return new GenericResponse(messages.getMessage("message.resetPasswordSuc", null, locale));
}
```

Unfortunately this spring I don't have time to figure this out. I must be preparing for my exam. If yn of you feel like helping me out about this issue, feel free ;-)

**Comment by BujakiAttila** [comment]
I have checked. This is still an issue. Password recovery still not working properly.

Jonas said, that the new passwod works well after restarting the application.

I think it is pretty sure, that the passwords are cached, and are not reloaded when it is updated in the DB through Hibernate ORM.

**Comment by BujakiAttila** [comment]
I think we should use the UserService class to update the password as described here:
[Password reset tutorial](http://www.baeldung.com/spring-security-registration-i-forgot-my-password)

Maybe something like this should be done instead of the current solution:
```java
@RequestMapping(value = "/user/savePassword", method = RequestMethod.POST)
@PreAuthorize("hasRole('READ_PRIVILEGE')")
@ResponseBody
public GenericResponse savePassword(Locale locale, @RequestParam("password") String password) {
    User user = (User) SecurityContextHolder.getContext().getAuthentication().getPrincipal();
    userService.changeUserPassword(user, password);
    return new GenericResponse(messages.getMessage("message.resetPasswordSuc", null, locale));
}
```
